metricbeat/module/openmetrics/collector/collector.go

```diff
@@ -29,14 +36,232 @@ const (
+			return nil, err
+		}
+		openmetrics, err := p.NewOpenMetricsClient(base)
+		if err != nil {
+			return nil, err
+		}
+
+		promEventsGen, err := genFactory(base)
+		if err != nil {
+			return nil, err
+		}
+
+		ms := &MetricSet{
+			BaseMetricSet:   base,
+			openmetrics:     openmetrics,
+			namespace:       namespace,
+			promEventsGen:   promEventsGen,
+			eventGenStarted: false,
+		}
+		// store host here to use it as a pointer when building `up` metric
+		ms.host = ms.Host()
+		ms.excludeMetrics, err = p.CompilePatternList(config.MetricsFilters.ExcludeMetrics)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to compile exclude patterns")
+		}
+		ms.includeMetrics, err = p.CompilePatternList(config.MetricsFilters.IncludeMetrics)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to compile include patterns")
+		}
+
+		return ms, nil
+	}
+}
+
+// Fetch fetches data and reports it
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
+	if !m.eventGenStarted {
+		m.promEventsGen.Start()
+		m.eventGenStarted = true
+	}
+
+	families, err := m.openmetrics.GetFamilies()
+	eventList := map[textparse.MetricType]map[string]common.MapStr{}
+	if err != nil {
+		// send up event only
+		families = append(families, m.upMetricFamily(0.0))
+
+		// set the error to report it after sending the up event
+		err = errors.Wrap(err, "unable to decode response from openmetrics endpoint")
+	} else {
+		// add up event to the list
+		families = append(families, m.upMetricFamily(1.0))
+	}
+
+	for _, family := range families {
+		if m.skipFamily(family) {
+			continue
+		}
+		promEvents := m.promEventsGen.GenerateOpenMetricsEvents(family)
+
+		for _, promEvent := range promEvents {
+			labelsHash := promEvent.LabelsHash()
+			if _, ok := eventList[promEvent.Type]; !ok {
+				eventList[promEvent.Type] = make(map[string]common.MapStr)
+			}
+			if _, ok := eventList[promEvent.Type][labelsHash]; !ok {
+				eventList[promEvent.Type][labelsHash] = common.MapStr{}
+
+				// Add default instance label if not already there
+				if exists, _ := promEvent.Labels.HasKey(upMetricInstanceLabel); !exists {
+					promEvent.Labels.Put(upMetricInstanceLabel, m.Host())
+				}
+				// Add default job label if not already there
+				if exists, _ := promEvent.Labels.HasKey("job"); !exists {
+					promEvent.Labels.Put("job", m.Module().Name())
+				}
+				// Add labels
+				if len(promEvent.Labels) > 0 {
+					eventList[promEvent.Type][labelsHash]["labels"] = promEvent.Labels
+				}
+			}
+
+			if promEvent.Type != "" {
+				eventList[promEvent.Type][labelsHash]["type"] = promEvent.Type
+			}
+			if promEvent.Unit != "" {
+				eventList[promEvent.Type][labelsHash]["unit"] = promEvent.Unit
+			}
+
+			if len(promEvent.Exemplars) > 0 {
+				eventList[promEvent.Type][labelsHash]["exemplar"] = promEvent.Exemplars
+			}
+			// Accumulate metrics in the event
+			eventList[promEvent.Type][labelsHash].DeepUpdate(promEvent.Data)
```

Comment:
So we accumulate metrics of the same type and of the same labels, right? Is it even possible that there is a metric with the same name but of different type?

Reply:
No, it's not possible to have same name and different types, textparse returns error during parsing.

Reply:
Reference - https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#name